src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -23,281 +24,342 @@ class dstype(enum.Enum):
+        high_range = sorted_idmaps[-1]['range_high'] + 100000000
+        return (low_range, high_range)
 
-        short_name = domain.get('idmap_domain_name')
-        short_name_verified = False
+    @private
+    async def remove_winbind_idmap_tdb(self):
+        sysdataset = (await self.middleware.call('systemdatast.config'))['basename']
+        ts = str(datetime.datetime.now(datetime.timezone.utc).timestamp())[:10]
+        await self.middleware.call('zfs.snapshot.create', {'dataset': f'{sysdataset}/samba4',
+                                                           'name': f'wb_cache-{ts}'})
+        try:
+            os.remove('/var/db/system/samba4/winbindd_idmap.tdb')
+        except Exception:
+            self.logger.debug("Failed to remove winbindd_idmap.tdb.", exc_info=True)
 
-        for d in configured_domains:
-            if d['id'] == id:
-                id_verified = True
-            if d['name'] == short_name:
-                short_name_verified = True
+    @accepts()
+    @job(lock='clear_idmap_cache')
+    async def clear_idmap_cache(self, job):
+        """
+        Stop samba, remove the winbindd_cache.tdb file, start samba, flush samba's cache.
+        This should be performed after finalizing idmap changes.
+        """
+        await self.middleware.call('service.stop', 'cifs')
 
-        if id is not None and not id_verified:
-            verrors.add('domain.id', f'Domain [{id}] does not exist.')
+        try:
+            os.remove('/var/db/system/samba4/winbindd_cache.tdb')
+        except Exception as e:
+            self.logger.debug("Failed to remove winbindd_cache.tdb: %s" % e)
 
-        if short_name is not None and not short_name_verified:
-            verrors.add('domain.idmap_domain_name', f'Domain [{short_name}] does not exist.')
+        await self.middleware.call('service.start', 'cifs')
+        gencache_flush = await run(['net', 'cache', 'flush'], check=False)
+        if gencache_flush.returncode != 0:
+            raise CallError(f'Attempt to flush gencache failed with error: {gencache_flush.stderr.decode().strip()}')
 
     @private
-    async def _common_validate(self, idmap_backend, data):
-        """
-        Common validation checks for all idmap backends.
-
-        1) Check for a high range that is lower than the low range.
+    async def autodiscover_trusted_domains(self):
+        smb = await self.middleware.call('smb.config')
 
-        2) Check for overlap with other configured idmap ranges.
+        ad_idmap_backend = (await self.query([('name', '=', 'DS_TYPE_ACTIVEDIRECTORY')], {'get': True}))['idmap_backend']
+        if ad_idmap_backend == IdmapBackend.AUTORID.name:
+            self.logger.trace('Skipping auto-generation of trusted domains due to AutoRID being enabled.')
+            return
 
-        In some circumstances overlap is permitted:
+        wbinfo = await run(['/usr/local/bin/wbinfo', '-m', '--verbose'], check=False)
```

Comment:
Should maybe use only "wbinfo" because of Linux?

Reply:
Good idea.